Description: Add support for FreeBSD 12
Author: Andrew Gallatin <gallatin@gmail.com>
Origin: upstream, https://github.com/HewlettPackard/netperf/commit/61003f8
--- a/configure.ac
+++ b/configure.ac
@@ -467,7 +467,7 @@
 #include <sys/socket.h>
 ]])
 		case "$host" in
-		*-*-freebsd[[7-9]].* | *-*-freebsd1[[0-1]].* )
+		*-*-freebsd[[7-9]].* | *-*-freebsd1[[0-2]].* )
 			# FreeBSD 7.x and later SCTP support doesn't need -lsctp.
 			;;
 		*)
@@ -699,7 +699,7 @@
 			enable_cpuutil="kstat - auto"
 			NETCPU_SOURCE="kstat"
 			;;
-                     *-*-freebsd[[4-9]].* | *-*-freebsd1[[0-1]].* | *-*-netbsd[[1-9]].* )
+                     *-*-freebsd[[4-9]].* | *-*-freebsd1[[0-2]].* | *-*-netbsd[[1-9]].* )
 			use_cpuutil=true
 			AC_DEFINE([USE_SYSCTL],,[Use MumbleBSD's sysctl interface to measure CPU util.])
 			enable_cpuutil="sysctl - auto"
