Description: Some additional FreeBSD fixes from Andrew Gallatin.
Author: raj <raj@5bbd99f3-5903-0410-b283-f1d88047b228>
Origin: upstream, https://github.com/HewlettPackard/netperf/commit/328fe3b
--- a/AUTHORS
+++ b/AUTHORS
@@ -107,6 +107,7 @@
 fixes to configure to recognize Solaris 11
 fixes to netcpu_procstat.c for later linux kernels
 workarounds to get Linux to report ENOBUFS on TX queue overflows
+FreeBSD fixes
 
 Mark Cooper
 pointing-out the need for -lresolv when compiling -DDO_DNS on RedHat
@@ -306,4 +307,4 @@
 
 Weijia Song - a fix inspiration for a NULL pointer problem.
 
-Gisle Vanem - some Windows compilation fixes
\ No newline at end of file
+Gisle Vanem - some Windows compilation fixes
--- a/configure.ac
+++ b/configure.ac
@@ -467,7 +467,7 @@
 #include <sys/socket.h>
 ]])
 		case "$host" in
-		*-*-freebsd[78].*)
+		*-*-freebsd[[7-9]].* | *-*-freebsd1[[0-1]].* )
 			# FreeBSD 7.x and later SCTP support doesn't need -lsctp.
 			;;
 		*)
@@ -699,7 +699,7 @@
 			enable_cpuutil="kstat - auto"
 			NETCPU_SOURCE="kstat"
 			;;
-                     *-*-freebsd[[4-8]].* | *-*-netbsd[[1-9]].* )
+                     *-*-freebsd[[4-9]].* | *-*-freebsd1[[0-1]].* | *-*-netbsd[[1-9]].* )
 			use_cpuutil=true
 			AC_DEFINE([USE_SYSCTL],,[Use MumbleBSD's sysctl interface to measure CPU util.])
 			enable_cpuutil="sysctl - auto"
--- a/src/dscp.c
+++ b/src/dscp.c
@@ -54,6 +54,9 @@
  */
 
 #if HAVE_NETINET_IN_SYSTM_H
+#if defined(__FreeBSD__)
+#include <sys/types.h>
+#endif
 #include <netinet/in_systm.h>
 #endif
 #if HAVE_NETINET_IP_H
